engine/src/juliabox/plugins/bucket_gs/impl_gs.py

```
@@ -0,0 +1,113 @@
+
+    @staticmethod
+    def connect_bucket(bucket):
+        if bucket not in JBoxGS.BUCKETS:
+            JBoxGS.BUCKETS[bucket] = JBoxGS.connect().buckets().get(bucket=bucket).execute()
+        return JBoxGS.BUCKETS[bucket]
+
+    @staticmethod
+    def _get_mime_type(local_file):
+        mime = MimeTypes()
+        url = urllib.pathname2url(local_file)
+        mime_type = mime.guess_type(url)
+        return mime_type[0]
+
+    @staticmethod
+    def push(bucket, local_file, metadata=None):
+        objconn = JBoxGS.connect().objects()
+        fh = open(local_file, "rb")
+        media = MediaIoBaseUpload(fh, JBoxGS._get_mime_type(local_file))
+        if metadata:
+            retval = objconn.insert(bucket=bucket, media_body=media,
+                                    name=os.path.basename(local_file),
+                                    body={"metadata": metadata}).execute()
+        else:
+            retval = objconn.insert(bucket=bucket, media_body=media,
+                                    name=os.path.basename(local_file)).execute()
+        fh.close()
+        return retval
+
+    @staticmethod
+    def pull(bucket, local_file, metadata_only=False):
+        objname = os.path.basename(local_file)
+        if metadata_only:
+            k = None
+            try:
+                k = JBoxGS.connect().objects().get(bucket=bucket,
+                                                   object=objname).execute()
+            except:
+                pass
+            return k
+        else:
+            req = JBoxGS.connect().objects().get_media(bucket=bucket,
+                                                       object=objname)
+
+            fh = open(local_file, "wb")
+            downloader = MediaIoBaseDownload(fh, req, chunksize=1024*1024)
+            done = False
+            while not done:
+                try:
+                    _, done = downloader.next_chunk()
+                except:
+                    break
```

should probably let the exception be raised, and close `fh` in the `finally` clause